File: 257-binary-tree-paths/binary-tree-paths.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution(object):
    def binaryTreePaths(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: List[str]
        """
        result = []

        def dfs(node, path):
            if not node:
                return

            if path:
                path = path + "->" + str(node.val)
            else:
                path = str(node.val)

            if not node.left and not node.right:
                result.append(path)
                return

            dfs(node.left, path)
            dfs(node.right, path)

        dfs(root, "")
        return result

    

File: 257-binary-tree-paths/binary-tree-paths.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<string> binaryTreePaths(TreeNode* root) {
        vector<string> result;

        function<void(TreeNode*, string)> dfs = [&] (TreeNode* node, string path){
            if (!node) return;

            if (path.empty()){
                path = to_string(node->val);
            }
            else{
                path += "->" + to_string(node->val);
            }
            if (!node->left && !node-> right){
                result.push_back(path);
                return;
            }
            dfs(node->left,path);
            dfs(node->right,path);
        };
        dfs(root,"");
        return result;
        
    }

};